src/Compilers/CSharp/Portable/Lowering/Instrumentation/LocalStateTracingInstrumenter.cs

```diff
@@ -0,0 +1,508 @@
+                    new[] { _factory.MethodDefIndex(_factory.TopLevelMethod), _factory.MethodDefIndex(_factory.CurrentFunction), _factory.StateMachineInstanceId() }),
+            };
+
+            var entryLogger = GetWellKnownMethodSymbol(entryOverload, _factory.Syntax);
+            var instrumentationPrologue = (entryLogger != null) ?
+                _factory.Assignment(_factory.Local(_scope.ContextVariable), _factory.Call(receiver: null, entryLogger, entryArgs)) : _factory.NoOp(NoOpStatementFlavor.Default);
+
+            var returnLogger = GetWellKnownMethodSymbol(WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogReturn, _factory.Syntax);
+            var instrumentationEpilogue = (returnLogger != null) ?
+                _factory.ExpressionStatement(_factory.Call(receiver: _factory.Local(_scope.ContextVariable), returnLogger)) : _factory.NoOp(NoOpStatementFlavor.Default);
+
+            // currently don't need to compose multiple instrumentations
+            Debug.Assert(instrumentation is null);
+
+            instrumentation = new BoundBlockInstrumentation(
+                _factory.Syntax,
+                _scope.ContextVariable,
+                instrumentationPrologue,
+                instrumentationEpilogue);
+
+            _scope.Close(isMethodBody);
+        }
+
+        public override BoundExpression InstrumentUserDefinedLocalAssignment(BoundAssignmentOperator original)
+        {
+            Debug.Assert(original.Left is BoundLocal { LocalSymbol.SynthesizedKind: SynthesizedLocalKind.UserDefined } or BoundParameter);
+
+            var assignment = base.InstrumentUserDefinedLocalAssignment(original);
+
+            bool? refAssignmentSourceIsLocal;
+            BoundExpression? refAssignmentSourceIndex;
+            if (original.IsRef)
+            {
+                if (original.Right is BoundLocal { LocalSymbol.SynthesizedKind: SynthesizedLocalKind.UserDefined } rightLocal)
+                {
+                    refAssignmentSourceIsLocal = true;
+                    refAssignmentSourceIndex = _factory.LocalId(rightLocal.LocalSymbol);
+                }
+                else if (original.Right is BoundParameter rightParameter)
+                {
+                    refAssignmentSourceIsLocal = false;
+                    refAssignmentSourceIndex = _factory.ParameterId(rightParameter.ParameterSymbol);
+                }
+                else
+                {
+                    // We don't track field aliases.
+                    return assignment;
+                }
+            }
+            else
+            {
+                refAssignmentSourceIsLocal = null;
+                refAssignmentSourceIndex = null;
+            }
+
+            var isLocalOrParameter = TryGetLocalOrParameterInfo(original.Left, out var targetSymbol, out var targetType, out var targetIndex);
+            Debug.Assert(isLocalOrParameter);
+            Debug.Assert(targetType is not null); // TODO: why are these needed?
```

Comment:
Found couple TODO comments in new code. Were you planning on resolving those before merging?

Reply:
These are questions for the compiler team :) IDK why this assert is needed. Without it the compiler complains about `targetType` being possibly null later on in this method.

Reply:
Am I missing anything or is this a bug? If the latter I'll file an issue.

Reply:
I see. Looks like a bug in nullability analysis to me.

Reply:
Filed issue https://github.com/dotnet/roslyn/issues/66807